apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+{
+    /// <summary>
+    /// Provides streaming access to a Spanner SQL query that automatically retries, handles
+    /// chunking and recoverable errors.
+    /// </summary>
+    public sealed class ReliableStreamReader : IDisposable
+    {
+        private AsyncServerStreamingCall<PartialResultSet> _currentCall;
+        private readonly object _sync = new object();
+        private readonly SpannerClient _spannerClient;
+        private readonly ExecuteSqlRequest _request;
+        private readonly Session _session;
+        private readonly CallSettings _callTiming;
+        private ResultSetMetadata _metadata;
+        private bool _isReading = true;
+
+        private ByteString _resumeToken;
+        private int _resumeSkipCount;
+
+        internal ReliableStreamReader(SpannerClient spannerClientImpl, ExecuteSqlRequest request, Session session)
+        {
+            _spannerClient = spannerClientImpl;
+            _request = request;
+            _session = session;
+            _clock = SpannerSettings.GetDefault().Clock ?? SystemClock.Instance;
+            _scheduler = SpannerSettings.GetDefault().Scheduler ?? SystemScheduler.Instance;
+
+            _callTiming = CallSettings.FromCallTiming(CallTiming.FromRetry(new RetrySettings(
+                    SpannerSettings.GetDefaultRetryBackoff(),
+                    SpannerSettings.GetDefaultTimeoutBackoff(),
+                    Expiration.FromTimeout(TimeSpan.FromMilliseconds(600000)),
+                    SpannerSettings.IdempotentRetryFilter
+                )));
+            _request.SessionAsSessionName = _session.SessionName;
+        }
+
+        private async Task<Metadata> ConnectAsync()
+        {
+            Logger.LogPerformanceCounterFn("StreamReader.ConnectCount", x => x + 1);
+            if (_resumeToken != null)
+            {
+                Logger.Debug(() => $"Resuming at location:{_resumeToken}");
+                _request.ResumeToken = _resumeToken;
+            }
+            _currentCall = _spannerClient.ExecuteSqlStream(_request);
+            return await _currentCall.ResponseHeadersAsync.ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Connects or reconnects to Spanner, fast forwarding to where we left off based on
+        /// our _resumeToken and _resumeSkipCount.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> ReliableConnect(CancellationToken cancellationToken)
+        {
+            if (_currentCall == null)
+            {
+//                Func<Task<Metadata>> connectFn = ConnectAsync;
+//                var responseHeaders = await connectFn.CallWithRetry(_callTiming, _clock, _scheduler);
+                await ConnectAsync().ConfigureAwait(false);
+
+                Debug.Assert(_currentCall != null, "_currentCall != null");
+                cancellationToken.ThrowIfCancellationRequested();
+
+                for (int i = 0; i <= _resumeSkipCount; i++)
```

Comment:
it probably wont be very large.  It certainly wont be larger than int.max.  You are correct that this would typically be seen where a large item is sent in several chunks and each of those chunks isn't assigned a resume token.

one thing we'll need to test is how well the api handles these large objects.  Right now, it loads an entire row, which may include one of these large objects since the api allows direct access to any column assuming a whole row is loaded.  So I'm not sure there is a way to stream a very large column.  However, I think if you store a 100MB file within spanner, you could argue that you are "doing it wrong".